### PR TITLE
mgr/dashboard: cluster upgrade start UI

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/61928
Signed-off-by: avanthakkar <avanjohn@gmail.com>

[Screencast from 2023-07-24 15-39-08.webm](https://github.com/ceph/ceph/assets/23611106/6c604d63-2807-4bf0-97da-a08b612de241)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
Some Nits:

Are we checking two times if the update is available? once when we load the upgrade page and once we are opening the modal for upgrade, can't we just pass data from parent component to child and use it?

Also set the default height and width of modal to size you think it will require.

**Comment by avanthakkar** [comment]
> Some Nits:
> 
> Are we checking two times if the update is available? once when we load the upgrade page and once we are opening the modal for upgrade, can't we just pass data from parent component to child and use it?
> 
> Also set the default height and width of modal to size you think it will require.

We need to subscribe again to see which versions are available. So probably I can update the message as 
`Fetching the available versions`? Not sure if I can @Input these versions. wdyt? @nizamial09 
Sure will do some resizing for text and modal.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  permission: Permission;
+  upgradeInfoError$: Observable<any>;
+  upgradeInfo$: Observable<UpgradeInfoInterface>;
+  upgradeForm: CdFormGroup;
+  icons = Icons;
+
+  constructor(
+    public actionLabels: ActionLabelsI18n,
+    private authStorageService: AuthStorageService,
+    private fb: CdFormBuilder,
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts:33]: don't need this
```suggestion
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    public activeModal: NgbActiveModal,
+    private upgradeService: UpgradeService,
+    private notificationService: NotificationService
+  ) {
+    this.permission = this.authStorageService.getPermissions().configOpt;
+  }
+
+  ngOnInit() {
+    this.upgradeInfo$ = this.upgradeService.list();
+    this.upgradeInfoError$ = this.upgradeInfo$?.pipe(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts:43]: we don't need to do this again. We can pass the versions from the upgrade.component where we are actually doing the subscription. so on this file you can introduce a variable called `versions` to catch the params that are send from upgrade component.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      ignoreElements(),
+      catchError((error) => of(error))
+    );
+
+    this.upgradeForm = this.fb.group({
+      availableVersions: this.fb.control(null, [Validators.required])
+    });
+  }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.ts:51]:
```suggestion
    this.upgradeForm = new CdFormGroup({
      availableVersions: new FormControl(null, [Validators.required])
    });
  }
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.ts [hunk 4]
```diff
@@ -46,4 +51,8 @@ export class UpgradeComponent implements OnInit {
     this.healthData$ = this.healthService.getMinimalHealth();
     this.fsid$ = this.healthService.getClusterFsid();
   }
+
+  startUpgradeModal() {
+    this.modalRef = this.modalService.show(UpgradeStartModalComponent);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade.component.ts:56]:
```suggestion
    this.modalRef = this.modalService.show(UpgradeStartModalComponent, {
        versions: this.upgradableVersions
    });
```

you have to get the `upgradableVersions` using a `pipe` from above where the `upgradeInfo$` is instanitated.

```
this.upgradeInfo$ = this.upgradeService.list().pipe(
      tap((upgradeInfo: UpgradeInfoInterface) => this.upgradableVersions = upgradeInfo.versions));
 
 ```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/constants/app.constants.ts [hunk 5]
```diff
@@ -215,6 +216,8 @@ export class ActionLabelsI18n {
     this.REMOVE_SCHEDULING = $localize`Remove Scheduling`;
     this.PROMOTE = $localize`Promote`;
     this.DEMOTE = $localize`Demote`;
+
+    this.UPGRADE = $localize`Start Upgrade`;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/constants/app.constants.ts:220]:
```suggestion
    this.START_UPGRADE = $localize`Start Upgrade`;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.html [hunk 6]
```diff
@@ -0,0 +1,60 @@
+<cd-modal [modalRef]="activeModal">
+  <ng-container class="modal-title">
+    <ng-container i18n>Upgrade Cluster</ng-container>&nbsp;
+  </ng-container>
+
+  <ng-container class="modal-content"
+                *ngIf="upgradeInfo$ | async as upgradeInfo; else checkUpgrade">
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.html:7]:
```suggestion
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.html [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                      [ngValue]="null"
+                      i18n>-- No version available --</option>
+              <option *ngIf="upgradeInfo.versions !== null && upgradeInfo.versions.length > 0"
+                      [ngValue]="null"
+                      i18n>-- Select a version --</option>
+              <option *ngFor="let version of upgradeInfo.versions"
+                      [value]="version">{{ version }}</option>
+            </select>
+            <span class="invalid-feedback"
+                  *ngIf="upgradeForm.showError('availableVersions', upgradeForm, 'required')"
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-start-modal-component.html:36]:
```suggestion
                  *ngIf="upgradeForm.showError('availableVersions', formDir, 'required')"
```

**Review by nizamial09 — CHANGES_REQUESTED** [review]
I've made some modifications in the review. If you follow that, you can avoid the duplicated api call. 

Also the `upgrade-start` component you created should really be `upgrade-form` component and it should be inside `upgrade/upgrade-form` folder rather than existing in the same folder as `upgrade`.

Also it is missing the unit-test file. You should also create that and push it.

**Comment by cloudbehl** [comment]
The version IMO needs to be in ascending order. 

image

**Comment by avanthakkar** [comment]
> The version IMO needs to be in ascending order.
> 
> image

Done!

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by cloudbehl** [comment]
Nit: `New versions` -> `New version`

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Review by nizamial09 — APPROVED** [review]

**Review by Pegonzal — APPROVED** [review]
Hey @avanthakkar , overall looks good to me and is working so I'll approve it!

But I agree with Nizam's comment which hasn't been taking into account yet, I think this should be done:

>
>Also the upgrade-start component you created should really be upgrade-form component and it should be inside upgrade/upgrade-form folder rather than existing in the same folder as upgrade.
>
>Also it is missing the unit-test file. You should also create that and push it.

**Comment by avanthakkar** [comment]
> Hey @avanthakkar , overall looks good to me and is working so I'll approve it!
> 
> But I agree with Nizam's comment which hasn't been taking into account yet, I think this should be done:
> 
> > Also the upgrade-start component you created should really be upgrade-form component and it should be inside upgrade/upgrade-form folder rather than existing in the same folder as upgrade.
> > Also it is missing the unit-test file. You should also create that and push it.

yeah missed that, will do unit test and move start modal component under `upgrade-form`

**Comment by avanthakkar** [comment]
> Nit: `New versions` -> `New version`

Done!

**Comment by avanthakkar** [comment]
> Hey @avanthakkar , overall looks good to me and is working so I'll approve it!
> 
> But I agree with Nizam's comment which hasn't been taking into account yet, I think this should be done:
> 
> > Also the upgrade-start component you created should really be upgrade-form component and it should be inside upgrade/upgrade-form folder rather than existing in the same folder as upgrade.
> > Also it is missing the unit-test file. You should also create that and push it.

Done!

**Comment by avanthakkar** [comment]
jenkins test make check

**Pull request merged**